crates/derive/src/sources/factory.rs

```diff
@@ -0,0 +1,68 @@
+//! Contains a Factory for creating a calldata and blob provider.
+
+use crate::sources::{BlobSource, CalldataSource, DataSource, PlasmaSource};
+use crate::traits::{ChainProvider, DataAvailabilityProvider};
+use crate::types::{BlockInfo, RollupConfig};
+use alloc::boxed::Box;
+use alloc::fmt::Debug;
+use alloy_primitives::Address;
+use anyhow::{anyhow, Result};
+use async_trait::async_trait;
+
+/// A factory for creating a calldata and blob provider.
+#[derive(Debug, Clone, Copy)]
+pub struct DataSourceFactory<CP>
+where
+    CP: ChainProvider + Clone,
+{
+    /// The chain provider to use for the factory.
+    pub chain_provider: CP,
+    /// The ecotone timestamp.
+    pub ecotone_timestamp: Option<u64>,
+    /// Whether or not plasma is enabled.
+    pub plasma_enabled: bool,
+    /// The L1 Signer.
+    pub signer: Address,
+}
+
+impl<F: ChainProvider + Clone> DataSourceFactory<F> {
+    /// Creates a new factory.
+    pub fn new(provider: F, cfg: RollupConfig) -> Self {
+        Self {
+            chain_provider: provider,
+            ecotone_timestamp: cfg.ecotone_time,
+            plasma_enabled: cfg.is_plasma_enabled(),
+            signer: cfg.l1_signer_address(),
+        }
+    }
+}
+
+#[async_trait]
+impl<F> DataAvailabilityProvider for DataSourceFactory<F>
+where
+    F: ChainProvider + Send + Sync + Clone + Debug,
+{
+    type DataIter = DataSource<F>;
```

Comment:
Recommend removing this associated type, and just returning a concrete `DataSource<F>` in `open_data`